src/test/sock_tests.cpp

```diff
@@ -0,0 +1,150 @@
+    // wronly pretend that the socket is not closed.
+    int type;
+    socklen_t len = sizeof(type);
+    return getsockopt(s, SOL_SOCKET, SO_TYPE, &type, &len) == -1 &&
+           (errno == EBADF || errno == ENOTSOCK);
+}
+
+static SOCKET CreateSocket()
+{
+    const int s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+    BOOST_REQUIRE(s != -1);
+    return static_cast<SOCKET>(s);
+}
+
+static void CreateSocketPair(int s[2])
+{
+    BOOST_REQUIRE_EQUAL(socketpair(AF_UNIX, SOCK_STREAM, 0, s), 0);
+}
+
+BOOST_AUTO_TEST_CASE(constructor_and_destructor)
+{
+    const SOCKET s = CreateSocket();
+    Sock* sock = new Sock(s);
+    BOOST_CHECK_EQUAL(sock->Get(), s);
+    BOOST_CHECK(!SocketIsClosed(s));
+    delete sock;
+    BOOST_CHECK(SocketIsClosed(s));
+}
+
+BOOST_AUTO_TEST_CASE(move_constructor)
+{
+    const SOCKET s = CreateSocket();
+    Sock* sock1 = new Sock(s);
+    Sock* sock2 = new Sock(std::move(*sock1));
+    delete sock1;
+    BOOST_CHECK(!SocketIsClosed(s));
+    BOOST_CHECK_EQUAL(sock2->Get(), s);
+    delete sock2;
+    BOOST_CHECK(SocketIsClosed(s));
+}
+
+BOOST_AUTO_TEST_CASE(move_assignment)
+{
+    const SOCKET s = CreateSocket();
+    Sock* sock1 = new Sock(s);
+    Sock* sock2 = new Sock();
+    *sock2 = std::move(*sock1);
+    delete sock1;
+    BOOST_CHECK(!SocketIsClosed(s));
+    BOOST_CHECK_EQUAL(sock2->Get(), s);
+    delete sock2;
+    BOOST_CHECK(SocketIsClosed(s));
+}
+
+BOOST_AUTO_TEST_CASE(release)
+{
+    const SOCKET s = CreateSocket();
+    Sock* sock = new Sock(s);
+    BOOST_CHECK_EQUAL(sock->Release(), s);
+    delete sock;
+    BOOST_CHECK(!SocketIsClosed(s));
+    BOOST_REQUIRE_EQUAL(close(s), 0);
+}
+
+BOOST_AUTO_TEST_CASE(reset)
+{
+    const SOCKET s = CreateSocket();
+    Sock sock(s);
+    sock.Reset();
+    BOOST_CHECK(SocketIsClosed(s));
+}
+
+static void SendAndRecvMessage(const Sock& sender, const Sock& receiver)
+{
+    const char* msg = "abcd";
+    constexpr size_t msg_len = 4;
+    char recv_buf[10];
+
+    BOOST_CHECK_EQUAL(sender.Send(msg, msg_len, 0), msg_len);
+    BOOST_CHECK_EQUAL(receiver.Recv(recv_buf, sizeof(recv_buf), 0), msg_len);
+    BOOST_CHECK_EQUAL(strncmp(msg, recv_buf, msg_len), 0);
+}
+
+BOOST_AUTO_TEST_CASE(send_and_receive)
+{
+    int s[2];
+    CreateSocketPair(s);
+
+    Sock* sock0 = new Sock(s[0]);
+    Sock* sock1 = new Sock(s[1]);
+
+    SendAndRecvMessage(*sock0, *sock1);
+
+    Sock* sock0moved = new Sock(std::move(*sock0));
+    Sock* sock1moved = new Sock();
+    *sock1moved = std::move(*sock1);
+
+    delete sock0;
+    delete sock1;
+
+    SendAndRecvMessage(*sock1moved, *sock0moved);
+
+    delete sock0moved;
+    delete sock1moved;
+
+    BOOST_CHECK(SocketIsClosed(s[0]));
+    BOOST_CHECK(SocketIsClosed(s[1]));
+}
+
+BOOST_AUTO_TEST_CASE(wait)
+{
+    int s[2];
+    CreateSocketPair(s);
+
+    Sock sock0(s[0]);
+    Sock sock1(s[1]);
+
+    std::thread waiter([&sock0]() { sock0.Wait(24h, Sock::RECV); });
```

Comment:
Does `24h` means 24 hours? If that's the case it looks too much to me, even for a UT.

Reply:
Yes, 24 hours. I wanted to use `1y` (1 year), but `y` is not supported.

If this test detects a failure, it would result in a "test timeout in CI" which waits for 2 hours.

I don't know how else to check if `Wait()` works as expected. Notice - it does not signal to the caller whether a timeout occurred or the socket is ready for the event. It was possible before, but [we simplified](https://github.com/bitcoin/bitcoin/pull/20788#discussion_r550250637) it because nobody of the callers used that information. I think it is not worth restoring it for the sake of this test, what's your take?

If that functionality is provided then the test could do: "wait 1ms; ensure it caused timeout, send some data to the socket, wait 1ms; ensure it did not timeout, but signaled that the socket is ready".

Reply:
> I think it is not worth restoring it for the sake of this test, what's your take?

I agree with you. I don't think it is a good idea to have code the sake of one test only.